synapse/storage/databases/main/media_repository.py

```diff
@@ -115,28 +115,54 @@ async def get_local_media(self, media_id: str) -> Optional[Dict[str, Any]]:
             desc="get_local_media",
         )
 
-    async def get_local_media_by_user(self, user_id: str) -> List[Dict[str, Any]]:
-        """Get the metadata for a local piece of media
-        which user_id has uploaded
+    async def get_local_media_by_user_paginate(
+        self, start: int, limit: int, user_id: str
+    ) -> Tuple[List[Dict[str, Any]], int]:
+        """Get a paginated list of metadata for a local piece of media
+        which an user_id has uploaded
 
+        Args:
+            start: offset in the list
+            limit: maximum amount of media_ids to retrieve
+            user_id: fully-qualified user id
         Returns:
-            A list of all metadata of user's media
+            A paginated list of all metadata of user's media
         """
 
-        return await self.db_pool.simple_select_list(
-            "local_media_repository",
-            {"user_id": user_id},
-            [
-                "media_id",
-                "media_type",
-                "media_length",
-                "upload_name",
-                "created_ts",
-                "last_access_ts",
-                "quarantined_by",
-                "safe_from_quarantine",
-            ],
-            "get_local_media_by_user",
+        def get_local_media_by_user_paginate_txn(txn):
+            sql_base = """
+                FROM local_media_repository
+                WHERE user_id = ?
+                """
+
+            args = [user_id]
+            sql = "SELECT COUNT(*) as total_media " + sql_base
+            txn.execute(sql, args)
+            count = txn.fetchone()[0]
+
+            sql = """
+                SELECT
+                    "media_id",
+                    "media_type",
+                    "media_length",
+                    "upload_name",
+                    "created_ts",
+                    "last_access_ts",
+                    "quarantined_by",
+                    "safe_from_quarantine"
+                {}
+                LIMIT ? OFFSET ?
```

Comment:
You need an `ORDER BY` clause here, otherwise the databases won't necessarily return the results in the same order each time. I'd suggest `ORDER BY created_ts DESC, media_id DESC`, which should give you the most recently uploaded media first (`media_id` is added so that `(user_id, created_ts, media_id)` is unique, which guarantees a complete ordering. Otherwise, in the unlikely case that the user uploaded two pieces of media in the same millisecond the ordering may swap them around).

Other APIs let this be configurable, but I don't think there is a need to do that here for now.